### PR TITLE
ARO-1423 and ARO-1422: Deny Editing ClusterVersion and UpgradeConfig

### PR DESCRIPTION
**Pull request by edisonLcardenas**
### Which issue this PR addresses:

[ARO-1422: UpgradeConfig](https://issues.redhat.com/browse/ARO-1422)
[ARO-1423: ClusterVersion](https://issues.redhat.com/browse/ARO-1423)

### What this PR does / why we need it:
Block editing of  ClusterVersion and UpgradeConfig by regular users. This will reduce amount of ticket we receive due to customers trying to do an unsupported action. We have a policy that will deny regulars users editing of ClusterVersion and UpgradeConfig.

### Test plan for issue:
* OPA Tests - **Passed**
* Gator Tests - **Passed**
* Real Cluster Verification - **Passed**
* Web Console Test - **Passed**

For detailed test result, please see the jiras above.

### Is there any documentation that needs to be updated for this PR?
Yes, the README for guardrails. Included  info on how to create policies that needs to sync data into OPA using data.inventory object.

### CODE REVIEW

File: pkg/operator/controllers/guardrails/policies/gktemplates/aro-deny-clusterversion.yaml [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+        	# ## If regular user and
+        	# ## has NO cloud.openshift.com entry in openshift-config/pull-secret Secret
+        	# ## ALLOW EDITING
+
+        	# ## If regular user and 
+        	# ## HAS cloud.openshift.com entry (`source: OCM` indicates pull-secret exists) in openshift-config/pull-secret Secret
+        	# ## NOT ALLOWED
+        	config_data := data.inventory.namespace["openshift-managed-upgrade-operator"]["v1"]["ConfigMap"]["managed-upgrade-operator-config"]["data"]["config.yaml"]
+        	regex.match("source: OCM", config_data)
```
Comment (ArrisLee) [inline pkg/operator/controllers/guardrails/policies/gktemplates/aro-deny-clusterversion.yaml:46]: ok, looks like this template is quite identical to the `aro-deny-upgradeconfig` one, so my understanding is, we need two different constraints to facilitate different matches, but can use reuse the template at least?

**Review by ArrisLee — CHANGES_REQUESTED** [review]
left a comment, I think we may be able to let these two constraints share the same template

File: pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-clusterversion-upgradeconfig/src.rego [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+	# ## If regular user and
+	# ## has NO cloud.openshift.com entry in openshift-config/pull-secret Secret
+	# ## ALLOW EDITING
+
+	# ## If regular user and 
+	# ## HAS cloud.openshift.com entry (`source: OCM` indicates pull-secret exists) in openshift-config/pull-secret Secret
+	# ## NOT ALLOWED
+	config_data := data.inventory.namespace["openshift-managed-upgrade-operator"]["v1"]["ConfigMap"]["managed-upgrade-operator-config"]["data"]["config.yaml"]
+	regex.match("source: OCM", config_data)
+	msg := "Modifying this resource is not allowed for regular users. This includes attempting to create, edit, or delete."
```
Comment (ArrisLee) [inline pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-clusterversion-upgradeconfig/src.rego:23]: I think we may need to provide a bit more info here, the condition here should be `MUO` is enabled right?

File: pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-clusterversion-upgradeconfig/src_test.rego [hunk 3]
```diff
@@ -0,0 +1,206 @@
+package arodenyclusterversionupgradeconfig
+
```
Comment (ArrisLee) [inline pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-clusterversion-upgradeconfig/src_test.rego:2]: I see you have combine `clusterversion` and `upgradeconfig` logics together as one constraint template, so do we need some more test cases here to cover `clusterversion` checks?

**Review by ArrisLee — CHANGES_REQUESTED** [review]
overall LGTM, left some comments and please fix lint warning

**Review by ArrisLee — APPROVED** [review]
LGTM

File: pkg/operator/controllers/guardrails/policies/gkconstraints/aro-clusterversion-upgradeconfig-deny.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+apiVersion: constraints.gatekeeper.sh/v1beta1
+kind: ARODenyClusterVersionUpgradeConfig
+metadata:
+  name: aro-deny-clusterversion-upgradeconfig-modification
```
Comment (yjst2012) [inline pkg/operator/controllers/guardrails/policies/gkconstraints/aro-clusterversion-upgradeconfig-deny.yaml:4]: we want to have the same .metadata.name as the filename of the constraint to keep feature flag name and constraint name consistent, so can you please rename it to aro-clusterversion-upgradeconfig-deny

File: pkg/operator/controllers/guardrails/policies/README.md [hunk 5]
```diff
@@ -136,17 +136,43 @@ spec:
         kinds: ["PodDisruptionBudget"]
 ```
 
-Make sure the filename of constraint is the same as the .metadata.name of the Constraint object, as it is the feature flag name that will be used to turn on / off the policy.
-
 ## Test Rego source code
 
+Make sure the filename of constraint is the same as the .metadata.name of the Constraint object, as it is the feature flag name that will be used to turn on / off the policy.
```
Comment (yjst2012) [inline pkg/operator/controllers/guardrails/policies/README.md:141]: why is the moving? feel like it is a note to the constraint section?

File: pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-clusterversion-upgradeconfig/sync.yaml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+          "openshift-operators",
+          "openshift-user-workload-monitoring",
+          "openshift-pipelines",
+          "openshift-marketplace",
+          "openshift-multus",
+          "openshift-network*",
+          "openshift-vsphere-*",
+          "openshift-config-*",
+          "openshift-console",
+          "openshift-service-ca*",
```
Comment (yjst2012) [inline pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-clusterversion-upgradeconfig/sync.yaml:30]: IIUC, as per [gk doc](https://open-policy-agent.github.io/gatekeeper/website/docs/exempt-namespaces/) this is excluding priv'd namespaces from gatekeeper's control, which will left them unprotected by our priv'd ns policy, isn't it?

File: pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-clusterversion-upgradeconfig/sync.yaml [hunk 7]
```diff
@@ -0,0 +1,40 @@
+apiVersion: config.gatekeeper.sh/v1alpha1
```
Comment (yjst2012) [inline pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-clusterversion-upgradeconfig/sync.yaml:1]: as to sync.yaml, need to move it to somewhere global to all policies and find a way to auto deploy it, we can put it here for now.

**Review by yjst2012 — CHANGES_REQUESTED** [review]
left a few comments, pls have a look, thanks.

File: pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-clusterversion-upgradeconfig/sync.yaml (hunk 6 above)
Comment (yjst2012) [reply pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-clusterversion-upgradeconfig/sync.yaml:30]: can you please test this out, by checking after this applied if the priv'd ns policy still work?

**Pull request merged**